### PR TITLE
Add sections to clarify dealing with inputs and outputs of a dynamic provider.

### PR DESCRIPTION
**Pull request by praneetloke**
This is how it looks when it is rendered.

https://user-images.githubusercontent.com/1466314/59140281-5b14f580-8950-11e9-9465-569e34b93d8a.png

### CODE REVIEW

File: content/reference/programming-model.md [hunk 1]
```diff
@@ -962,9 +962,93 @@ We can now create instances of the new `MyResource` resource kind in our program
...
 
+### Dynamic Resource Inputs
+
+The interface to other components outside of the dynamic provider is the `pulumi.dynamic.Resource`. To pass inputs to your implementation of the `pulumi.dynamic.ResourceProvider`, you actually need to pass them through the resource that is exposed to the outside world. That is the `props` parameter in the constructor in the example below. Any properties you set in the `props` object will be passed to your `pulumi.dynamic.ResourceProvider` functions as the `inputs` where appropriate.
```
Comment (pgavlin) [inline content/reference/programming-model.md]:
```suggestion
A dynamic provider interfaces with other components via subclasses of `pulumi.dynamic.Resource`. The inputs to a dynamic provider's implementation are provided via a instantiation of such a subclass; this is the `props` parameter in the constructor in the example below. Any properties you set in the `props` object will be passed to your `pulumi.dynamic.ResourceProvider` functions as the `inputs` where appropriate.
```

File: content/reference/programming-model.md [hunk 2]
```diff
@@ -962,9 +962,93 @@ We can now create instances of the new `MyResource` resource kind in our program
...
+```
+
+```go
+// Dynamic Providers are not supported in Go currently.
+```
+
 ### Resource Provider Interface
 
-Implementing the `pulumi.dynamic.ResourceProvider` interface requires implementing a subset of the methods below.  Each of these methods is asynchronous, and most common implementations of these methods will do asynchronous network I/O to provision resources in a backing cloud provider or other resource model. There are several important contracts to be aware of as part of determining how the Pulumi engine will call each of these methods, and with what data.
+Implementing the `pulumi.dynamic.ResourceProvider` interface requires implementing a subset of the methods below. Each of these methods is asynchronous, and most common implementations of these methods will do asynchronous network I/O to provision resources in a backing cloud provider or other resource model. There are several important contracts to be aware of as part of determining how the Pulumi engine will call each of these methods, and with what data.
```
Comment (pgavlin) [inline content/reference/programming-model.md]:
```suggestion
Implementing the `pulumi.dynamic.ResourceProvider` interface requires implementing a subset of the methods below. Each of these methods may be asynchronous, and most implementations of these methods will perform network I/O to provision resources in a backing cloud provider or other resource model. There are several important contracts between a dynamic provider and the Pulumi CLI that inform when these methods are called and with what data.
```

File: content/reference/programming-model.md [hunk 3]
```diff
@@ -962,9 +962,93 @@ We can now create instances of the new `MyResource` resource kind in our program
...
+
+Since the inputs to the custom resource are of type `pulumi.Input<T>`, when the functions below are called with those inputs they will be passed the resolved version of the same inputs. To get strong typing for your inputs in the provider functions, you should have another type that contains the same properties but with unwrapped-types as their types.
```
Comment (pgavlin) [inline content/reference/programming-model.md]:
```suggestion
Though the input properties passed to a `pulumi.dynamic.Resource` instance will usually be of type `pulumi.Input<T>` in order to compose well with other Pulumi resources, the dynamic provider's functions are invoked with the fully resolved input values. Strong typing for the inputs to your provider's functions can help clarify this; you can achieve this by creating a second interface with the same properties as your resource's inputs, but with fully unwrapped types.
```

File: content/reference/programming-model.md [hunk 4]
```diff
@@ -990,6 +1074,57 @@ Delete is invoked if the resource name (URN) exists in the previous state but no
...
+    public readonly myNumberOutput: pulumi.Output<number>;
+
+    constructor(name: string, props: MyResourceInputs, opts?: pulumi.CustomResourceOptions) {
+        super(myprovider, name, props, opts); 
+    }
+}
+```
+
+```python
+# Dynamic Providers are not supported in Python currently.
```
Comment (pgavlin) [inline content/reference/programming-model.md]:
```suggestion
# Dynamic Providers are not yet supported in Python.
```

File: content/reference/programming-model.md [hunk 5]
```diff
@@ -990,6 +1074,57 @@ Delete is invoked if the resource name (URN) exists in the previous state but no
...
+```
+
+```go
+// Dynamic Providers are not supported in Go currently.
```
Comment (pgavlin) [inline content/reference/programming-model.md]:
```suggestion
// Dynamic Providers are not yet supported in Go.
```

File: content/reference/programming-model.md (hunk 3 above)
Comment (praneetloke) [reply content/reference/programming-model.md]: I love this suggestion. Makes it read much better I think. Thanks!

**Comment by praneetloke** [comment]
@pgavlin 

> I am not sure that we need to deal with any language besides JS/TS in these docs. It is possible that dynamic providers may never come to other languages, as they do rely on a bit of magic that may be out of reach of those platforms.

Got it. I simply carried-through what was done in the existing sections under dynamic providers.

I have applied your suggestions.

**Comment by pgavlin** [comment]
> Got it. I simply carried-through what was done in the existing sections under dynamic providers.

SGTM. Consistency is definitely the right choice.

**Pull request merged**